asf_heat_pump_suitability/analysis/exploratory/create_full_dataset/assign_cluster_suitability_and_feasibility.py

```diff
@@ -0,0 +1,237 @@
+"""
+Assigns a suitable technology and computes feasibility scoring for each cluster.
+
+This file contains functions for:
+- categorising/assigning each cluster with the most suitable technology scheme.
+- computing the feasibility scores for different heating schemes and clusters;
+"""
+
+# package imports
+import polars as pl
+
+
+def assign_str_suitable_techology_scheme(
+    cluster_size: int,
+    hn_area: bool,
+    city_centre: bool,
+    has_outdoor_space: bool,
+) -> str:
+    """
+    Categorises a cluster (based on its size and other characteristics) into its
+    most suitable technology scheme:
+    - "individual_ashp": Individual air source heat pump (ASHP)
+    - "collective_ashp": Collective purchasing of ASHP
+    - "shared_ground_loop": Shared Ground Loop (SGL)
+    - "heat_network": Heat Network (HN)
+
+    Args:
+        cluster_size (int): The number of UPRNs in the cluster.
+        hn_area (bool): Indicates if the cluster is in a heat network area.
+        city_centre (bool): Indicates if the cluster is close to the city centre.
+        has_outdoor_space (bool): Indicates if the cluster has outdoor space.
+
+    Returns:
+        str: A string representing the suitability category of the cluster.
+    """
+
+    if cluster_size == 1:
+        return "individual_ashp"
+    elif hn_area:  # if the cluster is in a heat network area
+        return "heat_network"
+    else:  # not in a heat network area
+        if city_centre:
+            return "heat_network"
+        else:
+            if cluster_size > 20:
+                if has_outdoor_space:
+                    return "shared_ground_loop"
+                else:
+                    return "collective_ashp"
+            else:
+                return "collective_ashp"
+
+
+def create_df_suitability_categorisation(cluster_df: pl.DataFrame) -> pl.DataFrame:
+    """
+    Adds a new column to cluster_df with the suitability categorisation for each cluster.
+
+    Args:
+        cluster_df (pl.DataFrame): DataFrame containing information about each cluster of properties.
+
+    Returns:
+        pl.DataFrame: cluster_df with an additional column for the suitability categorisation.
+    """
+
+    cluster_df = cluster_df.with_columns(
+        pl.struct(["cluster_size", "hn_area", "city_centre", "has_outdoor_space"])
+        .map_elements(
+            lambda row: assign_str_suitable_techology_scheme(
+                row["cluster_size"],
+                row["hn_area"],
+                row["city_center"],
+                row["has_outdoor_space"],
+            )
+        )
+        .alias("suitable_scheme")
```

Comment:
```suggestion
        .alias("best_tech")
```
could we rename something along the lines of this? my suggestion isnt the best so open to other things but want something that indicates it's the top / best / selected tech and that it's a tech not a scheme

Reply:
I've called it "most_suitable_tech". Is that okay?